theano/gof/vm.py

```diff
@@ -890,6 +897,89 @@ def make_all(self, profiler=None, input_storage=None,
             compute_map[k] = [k.owner is None]
 
         thunks = []
+
+        # Collect Reallocation Info
+        compute_map_re = defaultdict(lambda: [0])
+        for var in fgraph.inputs:
+            compute_map_re[var][0] = 1
+
+        viewed_by = {}
+        for var in fgraph.variables:
+            viewed_by[var] = []
+        view_of = {}
+        reallocated_info = {}
+        dependencies = getattr(fgraph.profile, 'dependencies', {})
+        pre_allocated = set([])
+        allocated = set([])
+        for idx in range(len(order)):
+            node = order[idx]
+            dmap = getattr(node.op, 'destroy_map', None)
+            vmap = getattr(node.op, 'view_map', None)
+
+            idx_o = 0
+            for out in node.outputs:
+                for var in node.outputs:
+                    compute_map_re[var][0] = 1
+                ins = None
+                if dmap and idx_o in dmap:
+                    idx_v = dmap[idx_o]
+                    assert len(
+                        idx_v) == 1, "Here we only support the possibility to destroy one input"
+                    ins = node.inputs[idx_v[0]]
+                if vmap and idx_o in vmap:
+                    assert ins is None
+                    idx_v = vmap[idx_o]
+                    assert len(
+                        idx_v) == 1, "Here we only support the possibility to view one input"
+                    ins = node.inputs[idx_v[0]]
+                if ins is not None:
+                    assert isinstance(ins, theano.Variable)
+                    origin = view_of.get(ins, ins)
+                    view_of[out] = origin
+                    viewed_by[origin].append(out)
+                idx_o += 1
+
+            for ins in node.inputs:
+                assert not (ins in view_of and viewed_by[ins])
+                if (getattr(ins, 'ndim', None) == 0 and not storage_map[ins][0]
```

Comment:
Can you put as a comment that if the storage_map is not empty, it mean it is a constant or a shared variable? Can you double check for the shared variable? I'm not 100% sure of that case.